src/test/java/com/catapi/service/BreedServiceTest.java

```diff
@@ -0,0 +1,86 @@
+import com.catapi.view.BreedView;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestTemplate;
+import java.util.List;
+import java.util.Optional;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+class BreedServiceTest {
+    RestTemplate restTemplate = mock();
+    BreedRepository breedRepository = mock();
+    BreedService breedService = new BreedService(restTemplate, breedRepository);
+
+    @Test
+    void getAllBreedsFromExternalApi_should_return_list_of_BreedViews_if_body_of_responseEntity_is_not_null() {
+        List<BreedView> breedViewList = List.of(
+                new BreedView("1", "breedView 1", "origin 1", "description 1", "url 1"),
+                new BreedView("2", "breedView 2", "origin 2", "description 2", "url 2"),
+                new BreedView("3", "breedView 3", "origin 3", "description 3", "url 3"));
+
+        ResponseEntity<List<BreedView>> responseEntity = ResponseEntity.ok(breedViewList);
+        when(restTemplate.exchange(anyString(), any(), isNull(), any(ParameterizedTypeReference.class)))
+                .thenReturn(responseEntity);
+
+        List<BreedView> resultBreedViewList = breedService.getAllBreedsFromExternalApi();
+        List<BreedView> expected = List.of(
+                new BreedView("1", "breedView 1", "origin 1", "description 1", "url 1"),
+                new BreedView("2", "breedView 2", "origin 2", "description 2", "url 2"),
+                new BreedView("3", "breedView 3", "origin 3", "description 3", "url 3"));
+
+        assertEquals(expected, resultBreedViewList);
+    }
+
+    @Test
+    void getAllBreedsFromExternalApi_should_return_list_of_BreedViews_if_body_of_responseEntity_is_null() {
+        List<BreedView> breedViewList = List.of();
+        ResponseEntity<List<BreedView>> responseEntity = ResponseEntity.ok(breedViewList);
+
+        when(restTemplate.exchange(anyString(), any(), isNull(), any(ParameterizedTypeReference.class)))
+                .thenReturn(responseEntity);
+
+        List<BreedView> resultBreedViewList = breedService.getAllBreedsFromExternalApi();
+        List<BreedView> expected = List.of();
+
+        assertEquals(expected, resultBreedViewList);
+    }
+
+    @Test
+    void requestBreeds_should_throw_exception_with_right_message() {
+        when(restTemplate.exchange(anyString(), any(), isNull(), any(ParameterizedTypeReference.class)))
+                .thenThrow(new RuntimeException("Помилка"));
+        ExternalApiException e = assertThrows(ExternalApiException.class, () -> breedService.getAllBreedsFromExternalApi());
+        assertEquals("Cannot get breed from external api: Помилка", e.getMessage());
+    }
+
+
+    @Test
+    void updateBreedDb_should_update_existing_breeds_and_add_new_ones() {
```

Comment:
This test doesn't prove that it claims in its name.

It only shows that "save" method was called 2 times.

Reply:
Use this:

ArgumentCaptor<Breed> breedCaptor = ArgumentCaptor.forClass(Breed.class);
Mockito.verify(breedRepository, Mockito.times(2)).save(breedCaptor.capture());
List<Breed> arguments = breedCaptor.getAllValues();

this is the way how you can get actual arguments from method which is being called.

Then you can get arguments from list (in your case is List<Breeds>) and compare what you've got with what you expect

Reply:
how to compare? It was supposed that you will do something like that:

```
Breed expectedUpdated = new Breed();
expectedUpdated.setOuterBreedId(existingBreedEntity.getOuterBreedId());
 expectedUpdated.setBreedName(breedView1.name());
 expectedUpdated.setDescription(breedView1.description());
 assertEquals(expectedUpdated, savedBreeds.getFirst());
```

Reply:
Done